File: perf/benchmarks/micro/set_ops/set_ops.go
// set_ops — HashSet insert + membership micro-benchmark (Go).
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

func main() {
	n := int64(500_000)
	if len(os.Args) > 1 {
		v, _ := strconv.ParseInt(os.Args[1], 10, 64)
		n = v
	}
	t0 := time.Now()
	s := make(map[int64]struct{}, n)
	var seed int64 = 1
	mod := n * 2
	for i := int64(0); i < n; i++ {
		seed = (seed*1103515245 + 12345) & 0x7FFFFFFF
		s[seed%mod] = struct{}{}
	}
	var found int64 = 0
	for i := int64(0); i < n; i++ {
		seed = (seed*1103515245 + 12345) & 0x7FFFFFFF
		if _, ok := s[seed%mod]; ok {
			found++
		}
	}
	ms := float64(time.Since(t0).Nanoseconds()) / 1_000_000.0
	fmt.Printf("CHECKSUM:%d\n", found)
	fmt.Printf("ELAPSED_MS:%.6f\n", ms)
}

File: perf/benchmarks/micro/set_ops/set_ops.rs
// set_ops — HashSet insert + membership micro-benchmark (Rust).
use std::collections::HashSet;
use std::env;
use std::time::Instant;

fn main() {
    let args: Vec<String> = env::args().collect();
    let n: i64 = if args.len() > 1 { args[1].parse().unwrap_or(500_000) } else { 500_000 };
    let t0 = Instant::now();
    let mut s: HashSet<i64> = HashSet::with_capacity(n as usize);
    let mut seed: i64 = 1;
    let mod_v: i64 = n * 2;
    for _ in 0..n {
        seed = (seed.wrapping_mul(1103515245).wrapping_add(12345)) & 0x7FFF_FFFF;
        s.insert(seed % mod_v);
    }
    let mut found: i64 = 0;
    for _ in 0..n {
        seed = (seed.wrapping_mul(1103515245).wrapping_add(12345)) & 0x7FFF_FFFF;
        if s.contains(&(seed % mod_v)) {
            found += 1;
        }
    }
    let ms = t0.elapsed().as_nanos() as f64 / 1_000_000.0;
    println!("CHECKSUM:{}", found);
    println!("ELAPSED_MS:{:.6}", ms);
}
